lib/Index/IndexRecord.cpp

```diff
@@ -393,6 +396,195 @@ emitDataForSwiftSerializedModule(ModuleDecl *module,
                                  const PathRemapper &pathRemapper,
                                  SourceFile *initialFile);
 
+static void
+appendSymbolicInterfaceToIndexStorePath(SmallVectorImpl<char> &resultingPath) {
+  llvm::sys::path::append(resultingPath, "interfaces");
+}
+
+static bool initSymbolicInterfaceStorePath(StringRef storePath,
+                                           std::string &error) {
+  using namespace llvm::sys;
+  SmallString<128> subPath = storePath;
+  appendSymbolicInterfaceToIndexStorePath(subPath);
+  std::error_code ec = fs::create_directories(subPath);
+  if (ec) {
+    llvm::raw_string_ostream err(error);
+    err << "failed to create directory '" << subPath << "': " << ec.message();
+    return true;
+  }
+  return false;
+}
+
+static void appendSymbolicInterfaceClangModuleFilename(
+    StringRef filePath, SmallVectorImpl<char> &resultingPath) {
+  llvm::sys::path::append(resultingPath, llvm::sys::path::filename(filePath));
+  StringRef extension = ".symbolicswiftinterface";
+  resultingPath.append(extension.begin(), extension.end());
+}
+
+// FIXME (Alex): Share code with IndexUnitWriter in LLVM after refactoring it.
+static Optional<bool>
+isFileUpToDateForOutputFile(StringRef filePath,
+                            Optional<StringRef> timeCompareFilePath,
+                            std::string &error) {
```

Comment:
`llvm::ErrorOr<bool> isFileUpToDateForOutputFile(StringRef filePath, StringRef timeCompareFilePath)` instead? IMO I prefer an empty `StringRef` over `Optional<StringRef>`, but up to you.